### PR TITLE
Added flows for un-manage cluster

### PR DESCRIPTION
**Pull request by shtripat**
tendrl-bug-id: Tendrl/commons#797

Signed-off-by: Shubhendu <shtripat@redhat.com>

### CODE REVIEW

File: tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                for node_nw in nodes_nw.leaves:
+                    if node_nw in node_ids:
+                        NS._int.client.delete(
+                            "/networks/%s/%s" % (nw, node_nw),
+                            recursive=True
+                        )
+
+            # Load the gluster servers list and remove the cluster nodes
+            gl_srvr_list = NS._int.client.read(
+                "/indexes/tags/gluster/server"
```
Comment (r0h4n) [inline tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py]: This will also delete any gluster nodes from some other cluster in a multi cluster env. Please remove this

File: tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            for node_id in node_ids:
+                gl_srvr_list.remove(node_id)
+            NS._init.client.write(
+                "/indexes/tags/gluster/server",
+                gl_srvr_list
+            )
+
+            # Load the tendrl servers list and remove
+            # the cluster nodes
+            tendrl_nodes_list = NS._int.client.read(
```
Comment (r0h4n) [inline tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py]: why is this required

File: tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                tendrl_nodes_list.remove(node_id)
+            NS._init.client.write(
+                "/indexes/tags/tendrl/node",
+                tendrl_nodes_list
+            )
+
+            # Load the tendrl GI nodes list and remove
+            # the cluster nodes
+            tendrl_gi_nodes_list = NS._int.client.read(
+                "/indexes/tags/tendrl/integration/gluster"
```
Comment (r0h4n) [inline tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py]: why is this required

File: tendrl/commons/objects/definition/master.yaml [hunk 4]
```diff
@@ -111,6 +127,36 @@ namespace.tendrl:
       list: /clusters
       value: clusters/$TendrlContext.integration_id
       atoms:
+        StopNodeServices:
```
Comment (r0h4n) [inline tendrl/commons/objects/definition/master.yaml]: break this down into  StopMonitoringServices and StopIntegrationServices

File: tendrl/commons/objects/definition/master.yaml [hunk 5]
```diff
@@ -103,6 +103,22 @@ namespace.tendrl:
       type: Create
       uuid: 2f94a48a-05d7-408c-b400-e27827f4edef
       version: 1
+    UnmanageCluster:
+      tags:
+        - "tendrl/monitor"
+      atoms:
+        - tendrl.objects.Cluster.atoms.StopNodeServices
+        - tendrl.objects.Cluster.atoms.ClearClusterDetails
```
Comment (r0h4n) [inline tendrl/commons/objects/definition/master.yaml]: You can call this DeleteCluster

File: tendrl/commons/objects/definition/master.yaml [hunk 6]
```diff
@@ -103,6 +103,22 @@ namespace.tendrl:
...
+        - tendrl.objects.Cluster.atoms.ClearMonitoringData
```
Comment (r0h4n) [inline tendrl/commons/objects/definition/master.yaml]: DeleteMonitoring

File: tendrl/commons/objects/definition/master.yaml [hunk 7]
```diff
@@ -111,6 +127,36 @@ namespace.tendrl:
...
+          enabled: True
+          inputs:
+            mandatory:
+              - TendrlContext.integration_id
+          name: clear cluster details
+          run: tendrl.objects.Cluster.atoms.ClearClusterDetails
+          type: Update
+          uuid: 333c3333-3c33-33c3-333c-c33cc3c5555c
+          help: Clear cluster details
+        ClearMonitoringData:
```
Comment (r0h4n) [inline tendrl/commons/objects/definition/master.yaml]: DeleteMonitoring

File: tendrl/commons/objects/definition/master.yaml [hunk 8]
```diff
@@ -111,6 +127,36 @@ namespace.tendrl:
...
+          enabled: true
+          inputs:
+            mandatory:
+              - TendrlContext.integration_id
+          name: stop node services
+          run: tendrl.objects.Cluster.atoms.StopNodeServices
+          type: Update
+          uuid: 333c3333-3c33-33c3-333c-c33cc3c4444c
+          help: Stop node services
+        ClearClusterDetails:
```
Comment (r0h4n) [inline tendrl/commons/objects/definition/master.yaml]: DeleteCluster

File: tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
+import etcd
+
+
+from tendrl.commons.event import Event
+from tendrl.commons import objects
+from tendrl.commons.message import Message
+from tendrl.commons.objects.cluster_tendrl_context \
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py]: Done

File: tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    import ClusterTendrlContext
+from tendrl.commons.objects.node import Node
+
+
+class ClearClusterDetails(objects.BaseAtom):
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py]: Will change to DeleteClusterDetails

File: tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        super(ClearClusterDetails, self).__init__(*args, **kwargs)
+
+    def run(self):
+        integration_id = self.parameters['TendrlContext.integration_id']
+        cluster_tendrl_context = ClusterTendrlContext(
+            integration_id=integration_id
+        ).load()
+
+        try:
+            NS._int.client.delete(
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py]: Done

File: tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+                "/clusters/%s" % integration_id,
+                recursive=True
+            )
+            NS._int.client.delete(
+                "/alerting/clusters/%s" % integration_id,
+                recursive=True
+            )
+            NS._int.client.delete(
+                "/indexes/tags/detected_cluster_id_to_integration_id/%s" %
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py]: There are different data stored under `/indexes/tags` and they are based on node tags as well like `/indexes/tags/tendrl/monitor`. `/indexes/tags/tendrl/node` etc. How many such objects we create here?

File: tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+            nodes = NS._init.client.read(
+                "/clusters/%s/nodes" % integration_id
+            )
+            node_ids = []
+            for node_id in nodes.leaves:
+                node_ids.append(node_id)
+                nc = Node(node_id=node_id).load()
+                node_ips[node_id] = nc.fqdn
+                NS._int.client.delete(
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py]: so do you mean anything updated by tendrl-node-agent ideally shouldnt be deleted as part of un-manager? If so later as part of forget cluster, we would need to clear all these details. comments?

File: tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py (hunk 1 above)
Comment (shtripat) [reply tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py]: We are getting the list and remove only the nodes of the cluster from the list and then update. Below lines take care of the same.

File: tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py (hunk 2 above)
Comment (shtripat) [reply tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py]: If https://github.com/Tendrl/commons/pull/798#discussion_r160599756 is correct then yes this also can be removed.

File: tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py (hunk 3 above)
Comment (shtripat) [reply tendrl/commons/objects/cluster/atoms/clear_cluster_details/__init__.py]: If https://github.com/Tendrl/commons/pull/798#discussion_r160599756 is correct then yes this also can be removed.

File: tendrl/commons/objects/definition/master.yaml (hunk 5 above)
Comment (shtripat) [reply tendrl/commons/objects/definition/master.yaml]: DeleteClusterDetails

File: tendrl/commons/objects/definition/master.yaml (hunk 6 above)
Comment (shtripat) [reply tendrl/commons/objects/definition/master.yaml]: DeleteMonitoringDetails

File: tendrl/commons/objects/definition/master.yaml (hunk 4 above)
Comment (shtripat) [reply tendrl/commons/objects/definition/master.yaml]: Done

File: tendrl/commons/objects/definition/master.yaml (hunk 8 above)
Comment (shtripat) [reply tendrl/commons/objects/definition/master.yaml]: DeleteClusterDetails

File: tendrl/commons/objects/definition/master.yaml (hunk 7 above)
Comment (shtripat) [reply tendrl/commons/objects/definition/master.yaml]: DeleteMonitoringDetails

**Comment by shtripat** [comment]
Tested the latest change along with https://github.com/Tendrl/monitoring-integration/pull/317 and it works as expected. The cluster can be un-managed and managed back. Across un-manage and manage back, the integration-id for the cluster remains the same. Also while un-manager towards end there is a notification about cluster state change to `unmanaged` and location of the archived graphite data for the cluster.

@r0h4n still need to work on unittests for these changes. Will work out a separate PR for the same.

File: tendrl/commons/objects/cluster/atoms/stop_monitoring_services/__init__.py [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+            nodes = NS._int.client.read("/clusters/%s/nodes" % integration_id)
+            child_job_ids = []
+            node_ids = []
+            for node in nodes.leaves:
+                node_id = node.key.split("/")[-1]
+                node_ids.append(node_id)
+                # Create jobs on nodes for stoping services
+                _job_id = str(uuid.uuid4())
+                params = {
+                    "Services[]": ["collectd"]
```
Comment (r0h4n) [inline tendrl/commons/objects/cluster/atoms/stop_monitoring_services/__init__.py]: This needs to be documented, stating that Tendrl will have full control over "collectd" service on the storage nodes, which includes shutting it down

@julienlim @mbukatov @nthomas-redhat  whats your take on this

File: tendrl/commons/flows/unmanage_cluster/__init__.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+                    _cluster = NS.tendrl.objects.Cluster(
+                        integration_id=integration_id
+                    ).load()
+                    if _cluster.is_managed == "no":
+                        break
+                except etcd.EtcdKeyNotFound:
+                    time.sleep(5)
+                    continue
+            _cluster.status = ""
+            _cluster.current_job['status'] = "done"
```
Comment (r0h4n) [inline tendrl/commons/flows/unmanage_cluster/__init__.py]: we have been using "finished" all over Tendrl, lets stick to that instead of "done"

File: tendrl/commons/objects/cluster/atoms/configure_monitoring/__init__.py [hunk 16]
```diff
@@ -157,7 +157,8 @@ def run(self):
         _cluster = NS.tendrl.objects.Cluster(
             integration_id=NS.tendrl_context.integration_id
         ).load()
-        _cluster.import_status = "done"
+        _cluster.status = ""
+        _cluster.current_job['status'] = "done"
```
Comment (r0h4n) [inline tendrl/commons/objects/cluster/atoms/configure_monitoring/__init__.py]: we have been using "finished" all over Tendrl, lets stick to that instead of "done"

File: tendrl/commons/flows/unmanage_cluster/__init__.py [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+class UnmanageCluster(flows.BaseFlow):
+    def __init__(self, *args, **kwargs):
+        super(UnmanageCluster, self).__init__(*args, **kwargs)
+
+    def run(self):
+        integration_id = self.parameters['TendrlContext.integration_id']
+        _cluster = NS.tendrl.objects.Cluster(
+            integration_id=integration_id
+        ).load()
+        if _cluster.status is not None and \
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/flows/unmanage_cluster/__init__.py]: Ok will add cluster.is_managed as well

File: tendrl/commons/flows/unmanage_cluster/__init__.py [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+            _cluster.status != "" and \
+            _cluster.status in \
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/flows/unmanage_cluster/__init__.py]: ok. will try that

File: tendrl/commons/flows/unmanage_cluster/__init__.py (hunk 15 above)
Comment (shtripat) [reply tendrl/commons/flows/unmanage_cluster/__init__.py]: done

File: tendrl/commons/objects/cluster/atoms/configure_monitoring/__init__.py (hunk 16 above)
Comment (shtripat) [reply tendrl/commons/objects/cluster/atoms/configure_monitoring/__init__.py]: done

File: tendrl/commons/objects/cluster/atoms/delete_cluster_details/__init__.py [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+                    job_id=self.parameters['job_id'],
+                    flow_id=self.parameters['flow_id'],
+                )
+                continue
+
+        # Load the gluster servers list and remove
+        # the cluster nodes
+        try:
+            gl_srvr_list = NS._int.client.read(
+                "/indexes/tags/gluster/server"
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/objects/cluster/atoms/delete_cluster_details/__init__.py]: So actually I am not removing the tag actually here. I am just removing the nodes of the cluster being unmanaged from the list and write the tag back with pending node ids

File: tendrl/commons/objects/cluster/atoms/delete_cluster_details/__init__.py [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+            ).value
+            gl_srvr_list = ast.literal_eval(gl_srvr_list)
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/objects/cluster/atoms/delete_cluster_details/__init__.py]: done

File: tendrl/commons/objects/cluster/atoms/delete_cluster_details/__init__.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (r0h4n) [reply tendrl/commons/objects/cluster/atoms/delete_cluster_details/__init__.py]: I think we still need to improve this. read("/indexes/tags/gluster/server") gives you every gluster node irrespective of cluster id

[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/objects/cluster/atoms/delete_cluster_details/__init__.py]: If we remove the nodes of currently unmanaging cluster, it shouldnt matter for other clusters. How you see issue in this? Also what you suggest if not this way?

[the remark this answers is not in the document]
Comment (r0h4n) [reply tendrl/commons/objects/cluster/atoms/delete_cluster_details/__init__.py]: I would suggest we let the tendrl-gluster-integration clean up its own entry in /indexes/tags/gluster/server 

So when you shutdown the tendrl-gluster-integration service during unmanage you need to add this clean up of tags.

The reason I dont want to load /indexes/tags/gluster/server is because it adds overhead of going through each and every gluster node across all clusters

[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/objects/cluster/atoms/delete_cluster_details/__init__.py]: So you mean before stopping tendrl-gluster-integration, we should invoke a flow to clean the tags. So in gluster-integration as well we would need to load the list /indexes/tags/gluster/server and remove it node-id from the list and write it back.

So effectively each node loads this tag from etcd, removes its node-id from the list and writes it back. In existing case we are loading once in server and all the cluster nodes entries are removed from the list and then written back. Here is just one read and write whereas in other case (suggested one) it would `no os nodes * (1 read + 1 write)`

File: tendrl/commons/objects/cluster/atoms/delete_cluster_details/__init__.py [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+        etcd_keys_to_delete.append(
+            "/indexes/tags/detected_cluster/%s" %
+            cluster_tendrl_context.cluster_id
+        )
+        etcd_keys_to_delete.append(
+            "/indexes/tags/detected_cluster_id_to_integration_id/%s" %
+            cluster_tendrl_context.cluster_id
+        )
+        etcd_keys_to_delete.append(
+            "/indexes/tags/provisioner/%s" % integration_id
```
Comment (r0h4n) [inline tendrl/commons/objects/cluster/atoms/delete_cluster_details/__init__.py]: lets clean this tag in integration service

File: tendrl/commons/objects/cluster/atoms/delete_cluster_details/__init__.py [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+        )
+        etcd_keys_to_delete.append(
+            "/indexes/tags/tendrl/integration/%s" %
```
Comment (r0h4n) [inline tendrl/commons/objects/cluster/atoms/delete_cluster_details/__init__.py]: lets clean this tag in integration service

File: tendrl/commons/utils/event_utils.py [hunk 23]
```diff
@@ -25,8 +25,7 @@ def emit_event(resource, curr_value, msg, instance,
         plugin_instance=instance,
         message=msg,
         integration_id=integration_id or NS.tendrl_context.integration_id,
-        cluster_name=cluster_name or NS.tendrl_context.cluster_name,
-        sds_name=sds_name or NS.tendrl_context.sds_name,
+        cluster_name=cluster_name or NS.tendrl_context.cluster_name
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/utils/event_utils.py]: As confirmed by @GowthamShanmugam it was not required and sds_name internally is decided using integration_id.

File: tendrl/commons/objects/cluster/atoms/delete_cluster_details/__init__.py (hunk 21 above)
Comment (shtripat) [reply tendrl/commons/objects/cluster/atoms/delete_cluster_details/__init__.py]: Done

File: tendrl/commons/objects/cluster/atoms/delete_cluster_details/__init__.py (hunk 22 above)
Comment (shtripat) [reply tendrl/commons/objects/cluster/atoms/delete_cluster_details/__init__.py]: done

File: tendrl/commons/objects/definition/master.yaml [hunk 24]
```diff
@@ -103,6 +103,26 @@ namespace.tendrl:
...
+      enabled: true
+      inputs:
+        mandatory:
+          - TendrlContext.integration_id
+      pre_run:
+        - tendrl.objects.Cluster.atoms.IsClusterManaged
+      run: tendrl.flows.UnmanageCluster
+      type: Update
+      uuid: 2f94a48a-05d7-408c-b400-e27827f4efed
+      version: 1
```
Comment (nthomas-redhat) [inline tendrl/commons/objects/definition/master.yaml]: Do we need a post_run to make sure the cluster is un_managed and the services are stopped?

File: .travis.yml [hunk 25]
```diff
@@ -18,16 +18,18 @@ install:
...
 # All CI checks are executed via tox, one tox run for each env (see below).
 script:
   - tox
 
+# List of tox enviroments to run. This way, result for each env will be
+# reported separately.
+# See: http://borntyping.co.uk/2013/09/02/tox-and-travis/
 env:
-  - TOXENV=pep8
   - TOXENV=vulture
```
[the remark this answers is not in the document]
Comment (shtripat) [reply .travis.yml]: Done

File: tendrl/commons/flows/import_cluster/__init__.py [hunk 26]
```diff
@@ -13,22 +13,25 @@ def __init__(self, *args, **kwargs):
...
             integration_id = self.parameters['TendrlContext.integration_id']
-
             _cluster = NS.tendrl.objects.Cluster(
                 integration_id=NS.tendrl_context.integration_id).load()
-            if (_cluster.import_job_id is not None and
-                    _cluster.import_job_id != "") or _cluster.import_status \
-                    in ["in_progress", "done", "failed"]:
+            if (_cluster.status is not None and
+                    _cluster.status != "" and
+                    _cluster.status in ["syncing", "importing", "unmanaging"]):
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/flows/import_cluster/__init__.py]: Status field maintains status of any jobs running on the cluster. If import successful it would be set as syncing by GI. State of the cluster is maintained under final details as healthy /unhealthy. For afresh detected cluster status would be empty and I'd user submit import through app for a cluster which has import running back-end would reject the job.

File: tendrl/commons/flows/unmanage_cluster/__init__.py [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+            'status': "in_progress"
+        }
+        _cluster.save()
+
+        try:
+            super(UnmanageCluster, self).run()
+            _cluster = NS.tendrl.objects.Cluster(
+                integration_id=integration_id
+            ).load()
+            _cluster.status = ""
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/flows/unmanage_cluster/__init__.py]: State of cluster is maintained in global details object already and ui uses the same

File: tendrl/commons/objects/definition/master.yaml (hunk 24 above)
Comment (shtripat) [reply tendrl/commons/objects/definition/master.yaml]: May be a post task to make sure cluster is detected back again. Will check this part

File: tendrl/commons/flows/import_cluster/__init__.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (nthomas-redhat) [reply tendrl/commons/flows/import_cluster/__init__.py]: ack

[the remark this answers is not in the document]
Comment (nthomas-redhat) [reply tendrl/commons/flows/import_cluster/__init__.py]: ok

File: tendrl/commons/flows/unmanage_cluster/__init__.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (nthomas-redhat) [reply tendrl/commons/flows/unmanage_cluster/__init__.py]: ok

File: tendrl/commons/objects/definition/master.yaml (hunk 24 above)
Comment (nthomas-redhat) [reply tendrl/commons/objects/definition/master.yaml]: Please create an issue to track

Comment (shtripat) [reply tendrl/commons/objects/definition/master.yaml]: In case of node-agent re-start only we do sds_sync for cluster but in un-manage we are not (re)starting node-agent on storage nodes so removal of tag `indexes/tags/tendrl/integration/{int-id}` should not be done. Also we cannot depend on this to figure out that an un-managed cluster is import ready now.

@r0h4n any other suggestion??

Comment (shtripat) [reply tendrl/commons/objects/definition/master.yaml]: I have added a post task for the same and posted below

**Comment by julienlim** [comment]
Per discussion with @a2batic, there are a couple of UI changes related to Unmanage cluster.  We will be posting updates to the design soon.
